core/src/main/java/io/grpc/internal/Proxies.java

```diff
@@ -0,0 +1,138 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * A utility class to detect which proxy, if any, should be used for a given
+ * {@link InetSocketAddress}.
+ */
+public class Proxies {
+  private static final Logger log = Logger.getLogger(Proxies.class.getName());
+
+  @Deprecated
+  private static final String GRPC_PROXY_ENV_VAR = "GRPC_PROXY_EXP";
+
+  // We want an HTTPS proxy, which operates on the entire data stream (See IETF rfc2817).
+  private static final String URI_FORMAT = "https://%s";
+
+  /**
+   * Given a target address, returns which proxy address should be used. If no proxy should be
+   * used, then return null.
+   */
+  public static InetSocketAddress proxyFor(SocketAddress targetServerAddress) {
+    return proxyFor(
+        targetServerAddress,
+        ProxySelector.getDefault(),
+        System.getenv(GRPC_PROXY_ENV_VAR)
+    );
+  }
+
+  /**
+   * Helper method to achieve stateless unit tests.
+   */
+  @VisibleForTesting
+  static InetSocketAddress proxyFor(
+      SocketAddress targetServerAddress,
+      ProxySelector proxySelector,
+      @Nullable String proxyEnvStr
+  ) {
+    InetSocketAddress override = overrideProxy(proxyEnvStr);
+    if (override != null) {
+      return override;
+    }
+
+    return detectProxy(proxySelector, targetServerAddress);
+  }
+
+  /**
+   * GRPC_PROXY_EXP is deprecated but let's maintain compatibility for now.
+   */
+  private static InetSocketAddress overrideProxy(String proxyHostPort) {
+    if (proxyHostPort == null) {
+      return null;
+    }
+
+    String[] parts = proxyHostPort.split(":", 2);
+    int port = 80;
+    if (parts.length > 1) {
+      port = Integer.parseInt(parts[1]);
+    }
+    log.warning(
+        "Detected GRPC_PROXY_EXP and will honor it, but this feature will "
+            + "be removed in a future release. Use java.net.ProxySelector instead."
+    );
+    return new InetSocketAddress(parts[0], port);
+  }
+
+  private static InetSocketAddress detectProxy(
+      ProxySelector proxySelector,
+      SocketAddress targetServerAddress
+  ) {
+    if (!(targetServerAddress instanceof InetSocketAddress)) {
+      return null;
+    }
+
+    String hostName = ((InetSocketAddress) targetServerAddress).getHostName();
```

Comment:
If the socket address was created with an IP address this can do I/O. If you only expect this class to work with `DnsNameResolver`, you could `assert address.isUnresolved()`. But this problem has been causing some havoc. There is getHostString(), but it's not available in Java 6.